src/mpi/init/init_impl.c

```diff
@@ -107,17 +109,118 @@ const char **MPIR_pset_list = default_pset_list;
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/**
+ * @brief   Set the requested thread level of a session based on an info object
+ *
+ *          If the provided info_ptr is a nullptr or the thread_level key is not
+ *          found in the info object, the value of threadlevel remains unchanged.
+ *
+ * @param   info_ptr        Info object pointer to be parsed for thread_level key
+ * @param   threadlevel     Thread level found in info object (output)
+ *
+ * @return  int             MPI_SUCCESS or error code.
+ */
+static
+int get_thread_level_from_info(MPIR_Info * info_ptr, int *threadlevel)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int buflen = 0;
+    char *thread_level_s = NULL;
+    int flag = 0;
+    const char key[] = "thread_level";
+
+    /* No info pointer, nothing todo here */
+    if (info_ptr == NULL) {
+        goto fn_exit;
+    }
+
+    /* Get the length of the thread level value */
+    mpi_errno = MPIR_Info_get_valuelen_impl(info_ptr, key, &buflen, &flag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (!flag) {
+        /* Key thread_level not found in info object */
+        goto fn_exit;
+    }
+
+    /* Get thread level value. No need to check flag afterwards
+     * because we would not be here if thread_level key was not in info object.
+     */
+    thread_level_s = MPL_malloc(buflen + 1, MPL_MEM_BUFFER);
+    mpi_errno = MPIR_Info_get_impl(info_ptr, key, buflen, thread_level_s, &flag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Set requested thread level value as output */
+    mpi_errno = thread_level_to_int(thread_level_s, threadlevel);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    if (thread_level_s) {
+        MPL_free(thread_level_s);
+    }
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
 int MPIR_Session_init_impl(MPIR_Info * info_ptr, MPIR_Errhandler * errhandler_ptr,
                            MPIR_Session ** p_session_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
     MPIR_Session *session_ptr = NULL;
 
     int provided;
-    /* Let's try ask for MPI_THREAD_MULTIPLE, but it probably still works with
-     * MPI_THREAD_SINGLE since we use MPL_Initlock_lock for cross-session locks.
+    int requested = MPI_THREAD_MULTIPLE;
+
+    /* Get the thread level requested by the user via info object (if any) */
+    mpi_errno = get_thread_level_from_info(info_ptr, &requested);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Remark on MPI_THREAD_SINGLE: Multiple sessions may run in threads
+     * concurrently, so significant work is needed to support per-session MPI_THREAD_SINGLE.
+     * For now, it probably still works with MPI_THREAD_SINGLE since we use MPL_Initlock_lock
+     * for cross-session locks in MPII_Init_thread.
+     *
+     * TODO: support per-session MPI_THREAD_SINGLE and optimize
      */
-    mpi_errno = MPII_Init_thread(NULL, NULL, MPI_THREAD_MULTIPLE, &provided, &session_ptr);
+    mpi_errno = MPII_Init_thread(NULL, NULL, requested, &provided, &session_ptr);
```

Comment:
As I explained, we probably want to keep `MPI_THREAD_MULTIPLE` here. Otherwise, two threads using two separate sessions each using `MPI_THREAD_SERIALIZED` would race on MPICH's internal global resources. We need make sure `MPIR_ThreadInfo.isThreaded` is `true` for correctness.

Maybe we can provide a CVAR for users to "promise" us that they'll never concurrently use different sessions, to allow thread-single performance. Without such promise, I believe current code is incorrect.

Reply:
Changed the code back to hard-code `MPI_THREAD_MULTIPLE` here.

In addition, I've extended the comment before this line a bit. The standard recommends users to _not_ request `MPI_THREAD_SINGLE` for sessions to avoid complications.

I think further optimizations for `MPI_THREAD_SINGLE` cases should be done as separate PR. Here, the focus is on increasing the transparency of the `thread_level` parameter.